core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java

```diff
@@ -0,0 +1,128 @@
+
+import static com.linecorp.armeria.common.util.Functions.voidFunction;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.DefaultRpcResponse;
+import com.linecorp.armeria.common.RpcRequest;
+import com.linecorp.armeria.common.RpcResponse;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.channel.EventLoop;
+
+/**
+ * A {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+ */
+public class RetryingRpcClient extends RetryingClient<RpcRequest, RpcResponse> {
+    private static final RetryException RETRY_EXCEPTION = Exceptions.clearTrace(new RetryException());
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator() {
+        return delegate -> new RetryingRpcClient(delegate,
+                                                 RetryRequestStrategy.never(),
+                                                 Backoff::withoutDelay);
+    }
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator(RetryRequestStrategy<RpcRequest, RpcResponse> retryRequestStrategy) {
+        return delegate -> new RetryingRpcClient(delegate, retryRequestStrategy,
+                                                 Backoff::withoutDelay);
+    }
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator(RetryRequestStrategy<RpcRequest, RpcResponse> retryRequestStrategy,
+                 Supplier<? extends Backoff> backoffSupplier) {
+        return delegate -> new RetryingRpcClient(delegate, retryRequestStrategy, backoffSupplier);
+    }
+
+    /**
+     * Creates a new instance that decorates the specified {@link Client}.
+     */
+    public RetryingRpcClient(Client<? super RpcRequest, ? extends RpcResponse> delegate,
+                             RetryRequestStrategy<RpcRequest, RpcResponse> retryStrategy,
+                             Supplier<? extends Backoff> backoffSupplier) {
+        super(delegate, retryStrategy, backoffSupplier);
+    }
+
+    @Override
+    public RpcResponse execute(ClientRequestContext ctx, RpcRequest req) throws Exception {
+        DefaultRpcResponse responseFuture = new DefaultRpcResponse();
+        Backoff backoff = newBackoff();
+        retry(0, backoff, ctx, req, () -> {
```

Comment:
Shouldn't this be `1` and its parameter name be `currentAttemptNo` or something similar? Otherwise the `Backoff.nextIntervalMillis(...)` call in `retry()` below will not include the first attempt.